### ISSUE TITLE
"Improper nesting" logged out unexpectedly

### ISSUE THREAD

**Issue by osddeitf** [denoland/fresh#1773]
The issue exists for a while (if I remember correctly, it was introduced starting 1.2).
Whenever my component get rendered, there's a lot of weird output:
```
Add @babel/plugin-transform-react-jsx-source to get a more detailed component stack. Note that you should not add it to production builds of your App for bundle size reasons.
Improper nesting of table. Your <tr> should have a <thead/tbody/tfoot/table> parent.<tr class="suggestionRow">..</tr>

  in tr
  in SearchBar
  in 
  in Page

Improper nesting of table. Your <td> should have a <tr> parent.<td class="suggestionCategory" />

  in td
  in SearchBar
  in 
  in Page

Improper nesting of table. Your <td> should have a <tr> parent.<td class="suggestionList" />

  in td
  in SearchBar
  in 
  in Page
```
My component (as an island):
```jsx
<div
  class="inputSuggestions"
  style={showSuggestion && results.length ? { display: 'block' }: {}}>
  <table>
    <tr class='suggestionRow'>
      <td class='suggestionCategory'>
        <div class='suggestionDic'>English</div>
      </td>
      <td class='suggestionList'>
        {/*<ul></ul>*/}
      </td>
    </tr>
  </table>
</div>
```
My fresh version: v1.4.3.

**Comment by marvinhagemeister** [comment]
Whilst this is valid HTML, the browser will rewrite the structure to include a `<tbody>` element on the fly. When you take a look into the browser's DevTools you'll see that it has been rewritten to this:

```html
<table>
  <tbody>
    <tr class="suggestionRow">
      <td class="suggestionCategory">
        <div class="suggestionDic">English</div>
      </td>
      <td class="suggestionList">
      </td>
    </tr>
  </tbody>
</table>
```

This again messes with the expectations of any virtual-dom based framework like Preact and React where the elements it told the browser to render don't match up with what the browser did.

To resolve this, add the `<tbody>` tag to your table.

**Comment by osddeitf** [comment]
I did tried it, but still the same. You can see the message logged `Your <tr> should have a <thead/tbody/tfoot/table> parent`, it stated that either `<table>` or `<tbody>` is needed.

**Comment by osddeitf** [comment]
Also, browser doesn't add `<tbody>`, so it is not some kind of hydration mismatch:

![image](https://github.com/denoland/fresh/assets/37999210/80b8cca5-f6a0-4ba6-a059-07bc2a232a09)

**Comment by osddeitf** [comment]
By the look of the component stacks printed:
```
in tr
in SearchBar
```
and 
```
in td
in SearchBar
```
I think the issue is related to this fix: https://github.com/preactjs/preact/pull/4043

**Comment by marvinhagemeister** [comment]
@osddeitf Which browser is that? It adds the `tbody` element for me in Chrome. Maybe it's best if you share a reproduction case.

**Comment by osddeitf** [comment]
I was using Microsoft Edge, but it is not relevant now that I have made a minimal reproduction. You can create an island like this:
```
export default function BarSearch() {
  return (
    <div
      class="inputSuggestions"
      // style={showSuggestion && results.length ? { display: 'block' }: {}}>
      >
      <table>
        <tr class='suggestionRow'>
          <td class='suggestionCategory'>
            <div class='suggestionDic'>English</div>
          </td>
          <td class='suggestionList'>
            {/*<ul></ul>*/}
          </td>
        </tr>
      </table>
    </div>
  )
}
```
And even when using `curl` to fetch the route (which renders the island), the output will appear.

**Comment by osddeitf** [comment]
Not to mention the output also includes `Your <td> should have a <tr> parent`, that is not related to whether or not browsers add `<tbody>`.

**Comment by osddeitf** [comment]
@marvinhagemeister 
1. `preact/debug` saves the list of VNode(s) in options._diff hook: https://github.com/preactjs/preact/blob/10.17.1/debug/src/debug.js#L142
2. And check the list in `options._commit` hook: https://github.com/preactjs/preact/blob/10.17.1/debug/src/debug.js#L380
3. I debug and found out that `vnode._parent` is `undefined` for every `vnode`, then "Improper nesting..." is logged.
4. The problem lies in `preact-render-to-string`:
  - 4.1. When each `vnode` is being rendered, it is assigned a `vnode._parent` and `options._diff` is called: https://github.com/preactjs/preact-render-to-string/blob/v6.2.1/src/index.js#L173
  - 4.2. After each `vnode` rendered, its `vnode._parent` is cleared: https://github.com/preactjs/preact-render-to-string/blob/v6.2.1/src/index.js#L453
  - 4.3. And `options._commit` is called after the root `vnode` rendered: https://github.com/preactjs/preact-render-to-string/blob/v6.2.1/src/index.js#L68

So my proposal is we delete the step 4.2 (I tried and it did work), if that's alright for you, I'll create a PR for this.

**Comment by cknight** [comment]
I upgraded to Fresh 1.5.1 from 1.4.3 today and now get this error when running locally (via `deno task start`).  I've reduced the code to the example below. The error messages I get with this simple layout are:

```sh
Improper nesting of table. Your <thead/tbody/tfoot> should have a <table> parent.<thead />

  in thead
  in SearchResults
  in 
  in Search
  in 
  in Component

Improper nesting of table. Your <tr> should have a <thead/tbody/tfoot/table> parent.<tr>..</tr>

  in tr
  in SearchResults
  in 
  in Search
  in 
  in Component

Improper nesting of table. Your <th> should have a <tr>.<th>..</th>

  in th
  in SearchResults
  in 
  in Search
  in 
  in Component

Improper nesting of table. Your <th> should have a <tr>.<th>..</th>

  in th
  in SearchResults
  in 
  in Search
  in 
  in Component

Improper nesting of table. Your <th> should have a <tr>.<th>..</th>

  in th
  in SearchResults
  in 
  in Search
  in 
  in Component

Improper nesting of table. Your <thead/tbody/tfoot> should have a <table> parent.<tbody />

  in tbody
  in SearchResults
  in 
  in Search
  in 
  in Component

Improper nesting of table. Your <tr> should have a <thead/tbody/tfoot/table> parent.<tr>..</tr>

  in tr
  in SearchResults
  in 
  in Search
  in 
  in Component

Improper nesting of table. Your <td> should have a <tr> parent.<td id="1">..</td>

  in td
  in SearchResults
  in 
  in Search
  in 
  in Component

Improper nesting of table. Your <td> should have a <tr> parent.<td id="2">..</td>

  in td
  in SearchResults
  in 
  in Search
  in 
  in Component

Improper nesting of table. Your <td> should have a <tr> parent.<td id="3">..</td>

  in td
  in SearchResults
  in 
  in Search
  in 
  in Component
```
The relevant code in question from `SearchResults` is:
```tsx
return (
    <div>
      <table>
        <thead>
          <tr>
            <th>input box</th>
            <th>key</th>
            <th>value</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td id="1">a</td>
            <td id="2">b</td>
            <td id="3">c</td>
          </tr>
        </tbody>
      </table>
    </div>
  );
```

And the output sent to the browser is, e.g.
```html
<div>
  <table>
      <thead>
          <tr>
              <th>input box</th>
              <th>key</th>
              <th>value</th>
          </tr>
      </thead>
      <tbody>
          <tr>
              <td id="1">a</td>
              <td id="2">b</td>
              <td id="3">c</td>
          </tr>
      </tbody>
  </table>
</div>
```

All of which looks correct to me?  The only "special" bit of the `SearchResults` island is that it is nested inside another island.  No idea if that complicates things.  E.g. 
```tsx
  return (
    <>
        <SearchForm formIds={formIds}>
            <SearchBox/>
            <SearchResults/>
        </SearchForm>
    </>
  );
```

**Comment by cknight** [comment]
If anyone else is experiencing this and wishes (until the issue is resolved) to suppress the extremely verbose error messages which flood the console, you can achieve this with the following code (in dev.ts or main.ts):

```ts
const origConsoleError = console.error;
console.error = (msg) => {
  if (typeof msg === "string" && msg.includes("Improper nesting of table")) return;
  origConsoleError(msg);
};
```

**Comment by marvinhagemeister** [comment]
Re-opening until we've cut a new Preact release.

**Comment by kellnerd** [comment]
I also had this issue after upgrading from Fresh 1.2 to 1.5.4 and upgrading Preact in my import map made the console spam disappear again:
```diff
-	"preact": "https://esm.sh/preact@10.18.1",
-	"preact/": "https://esm.sh/preact@10.18.1/",
+	"preact": "https://esm.sh/preact@10.18.2",
+	"preact/": "https://esm.sh/preact@10.18.2/",
```
So I guess #2035 (upgrade to preact@10.19.1) has already fixed this in `main`.

**Comment by osddeitf** [comment]
I upgraded preact and this is fixed, so I'm closing this issue.

**Comment by marvinhagemeister** [comment]
Reopening until fix is released, in the meantime using a newer Preact version resolves this.

**Comment by marvinhagemeister** [comment]
We'll cut a patch release shortly. Resolved by #2038
